File: group-anagrams-49/group_anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    if(strs.length==0) {
        return [];
    }

    let _map = {};

    for(const str of strs) {
        let letters = new Array(26).fill(0);
        for(const c of str) {
            let index = c.charCodeAt() - 'a'.charCodeAt();
            letters[index]+=1;
        }
        let key = letters.toString();
        if(key in _map) {
            _map[key].push(str);
        }else {
            _map[key] = [str];
        }

        
    }

    return Object.values(_map);
};

File: group-anagrams-49/group_anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        if(strs.empty()) {
            return {};
        }

        unordered_map<string,vector<string>> map;

        for(int i=0; i< strs.size(); i++) {
            vector<int> letter(26,0);
            for(auto ch: strs[i]) {
                int index = int(ch) - int('a');
                letter[index]+=1;

            }

            string key;
            for(int count: letter) {
                key += "#" + to_string(count);
            }

            map[key].push_back(strs[i]);



        }

        vector<vector<string>> result;
        for (const auto& group : map) {
            result.push_back(group.second);
        }

        return result;
    }
};